packages/ipfs-core/src/components/dag/put.js

```diff
@@ -100,15 +66,93 @@ module.exports = ({ ipld, pin, gcLock, preload }) => {
 }
 
 /**
- * @typedef {Object} PutOptions
+ *
+ * @param {PutOptions} options
+ */
+const readEncodingOptions = (options) => {
+  if (options.cid && (options.format || options.hashAlg)) {
+    throw new Error('Can\'t put dag node. Please provide either `cid` OR `format` and `hashAlg` options.')
+  } else if (((options.format && !options.hashAlg) || (!options.format && options.hashAlg))) {
+    throw new Error('Can\'t put dag node. Please provide `format` AND `hashAlg` options.')
+  }
+
+  const { hashAlg, format } = options.cid != null
```

Comment:
Well so what happened is that this pull request added bunch of typedefs that were not there and typechecker then started complaining about bunch of issues that before it was not because it treated them as any. Specifically before this it had no idea what `config.ipld` was so it did not care. Now that it understands it, it's a lot pickier.

This will be simplified once I get around to #3349